index.coffee

```diff
@@ -204,83 +233,174 @@ exports.DockerProgress = class DockerProgress
+						extractSize[shortId] = layerSizes[remoteId]
+					else if status is 'Already exists'
+						remoteId = layerIds[shortId]
+						downloadSize[shortId] = layerSizes[remoteId]
+						extractSize[shortId] = layerSizes[remoteId]
+
+					if status.match(/^Status: Image is up to date for /)
+						downloadedSize = totalSize
+						extractedSize = totalSize
+					else
+						downloadedSize = _.sum(_.values(downloadSize))
+						extractedSize = _.sum(_.values(extractSize))
 
-					downloadedSize = completedSize + _.sum(_.values(currentDownloadedSize))
-					percentage = calculatePercentage(downloadedSize, totalSize)
+					completedSize = downloadedSize + extractedSize
+					downloadedPercentage = calculatePercentage(downloadedSize, totalSize)
+					extractedPercentage = calculatePercentage(extractedSize, totalSize)
+					percentage = calculatePercentage(completedSize, total)
 
-					onProgress(_.merge(evt, { downloadedSize, totalSize, percentage }))
+					onProgress _.merge evt, {
+						percentage
+						downloadedSize
+						downloadedPercentage
+						extractedSize
+						extractedPercentage
+						totalSize
+						totalProgress: renderProgress(percentage, completedSize, total)
+						totalProgressDetail:
+							current: completedSize
+							total: total
+					}
 				catch err
 					console.warn('Progress error:', err.message ? err)
-					totalSize = null
 
-	# Create a stream that transforms `docker.modem.followProgress` onProgress events to include total progress metrics.
+	# Get size of all layers of a local image
+	# "image" is a string, the name of the docker image
+	getImageLayerSizes: (image) ->
+		d = @docker
+		d.getImage(image).historyAsync()
+		.then (layers) ->
+			# roll-up sizes of layers with ID <missing> onto the parent.
+			# for example, transforms this:
+			#
+			# [{"Id": "sha256:a145e85e1...", "Size": 0, ...},
+			#  {"Id": "<missing>", "Size": 207, ...},
+			#  {"Id": "<missing>", "Size": 2125, ...}]
+			#
+			# to this: [{"Id": "sha256:a145e85e1...", "Size": 2332, ...}]
+			concreteLayers = []
+			accumulatedSize = 0
+			_.forEachRight layers, (layer) ->
+				if layer.Id is '<missing>'
+					accumulatedSize += layer.Size
+				else
+					layer.Size += accumulatedSize
+					accumulatedSize = 0
+					concreteLayers.push(layer)
+			return concreteLayers
+		.map (layer) ->
+			# transform into something like this:
+			# [ {id: "bb12b...", size: 1339, blobs: [ "a694d3...", ... ]}, ... ]
+			d.getImage(layer['Id']).inspectAsync().then (info) ->
+				id: stripShaPrefix(layer['Id'])
+				size: layer['Size']
+				blobs: _(info['RootFS']['Layers'])
+					.map(stripShaPrefix)
+					.uniq()
+					.value()
+		.then (layers) ->
+			# each layer references all blobs it's made of. for each layer
+			# keep only layer-specific blobs, dropping older ones.
+			knownBlobs = null
+			_.forEach layers, (layer) ->
+				if not knownBlobs?
+					knownBlobs = layer.blobs
+					return
+				layer.blobs = _.dropWhile layer.blobs, (blob) ->
```

Comment:
Hmm, you're right, but I now see the original code is wrong in the first place. Images can relate to several instances of the same blob because blobs IDs are their content's digest. This means `_.difference` (or `_.dropWhile`+`_.include`) will even remove newly appended blobs if these blobs have the same digest as existing ones. I should be comparing the two blob lists by index and break on the first differing one instead.